### PR TITLE
feat: add support for cloud artifacts; massive scraper refactor

### PR DESCRIPTION
**Pull request by dejanzele**
## Pull request description 

Added support for sending artifacts to cloud
Massive refactor of Scraper using Extractor -> Loader piping
Added unit and integration tests for Scraper

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [x] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: pkg/executor/scraper/filesystem_extractor.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+				if fileInfo.IsDir() {
+					return nil
+				}
+
+				reader, err := e.fs.OpenFileBuffered(path)
+				if err != nil {
+					return errors.Wrapf(err, "error opening buffered %s", path)
+				}
+				object := &Object{
```
Comment (exu) [inline pkg/executor/scraper/filesystem_extractor.go:38]: Why not File - Object is quite generic name

File: pkg/executor/scraper/loader.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package scraper
+
+import (
+	"context"
+)
+
+//go:generate mockgen -destination=./mock_loader.go -package=scraper "github.com/kubeshop/testkube/pkg/executor/scraper" Loader
+type Loader interface {
```
Comment (exu) [inline pkg/executor/scraper/loader.go:8]: Isn't  it `Uploader`? As it's not loading anything :)

File: pkg/executor/scraper/minio_loader.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	err := client.Connect()
+	if err != nil {
+		return nil, errors.Errorf("error occured creating minio client: %v", err)
+	}
+	l.client = client
+
+	return l, nil
+}
+
+func (l *MinIOLoader) Load(ctx context.Context, object *Object, meta map[string]any) error {
```
Comment (exu) [inline pkg/executor/scraper/minio_loader.go:41]: Looks like `Load` just Store/Save/Uploads file with `client.SaveFileDirect(` so Load is confusing

File: pkg/executor/scraper/scraper.go [hunk 4]
```diff
@@ -14,43 +10,22 @@ type Scraper interface {
...
-		Endpoint:        endpoint,
-		AccessKeyID:     accessKeyID,
-		SecretAccessKey: secretAccessKey,
-		Location:        location,
-		Token:           token,
-		Bucket:          bucket,
-		Ssl:             ssl,
-	}
-
+type ScraperV2 struct {
```
Comment (exu) [inline pkg/executor/scraper/scraper.go:13]: I think we should discuss versioning conventions (V2) usually in go there is subdirectory for V2

And do we need V2 here? As looks like you're overriding existing one

In This example I don't think we need V2 at all
Just from Testkube version 1.10 (some minor bump) we'll allow to handle Cloud + MinIO

**Review by exu — APPROVED** [review]

File: pkg/executor/scraper/scraper.go (hunk 4 above)
Comment (vsukhin) [reply pkg/executor/scraper/scraper.go:13]: If we introduce V2, then we need to have V1 as well. If you just refactor the package, you can keep existing version

File: pkg/executor/scraper/minio_loader.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	err := client.Connect()
+	if err != nil {
+		return errors.Errorf("error occured creating minio client: %v", err)
+	}
+
+	if err := utils.CheckStringKey(meta, "executionId"); err != nil {
+		return err
+	}
+
+	folder := meta["executionId"].(string)
```
Comment (vsukhin) [inline pkg/executor/scraper/minio_loader.go:52]: possible panic

Comment (vsukhin) [reply pkg/executor/scraper/minio_loader.go:52]: sounds like, it should be GetStringKey in order to not have this statement

File: pkg/cloud/data/scraper/loader.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		return nil, err
+	}
+	if err := utils.CheckStringKey(info, "executionId"); err != nil {
+		return nil, err
+	}
+	if err := utils.CheckStringKey(info, "testName"); err != nil {
+		return nil, err
+	}
+	req := &PutObjectSignedURLRequest{
+		Object:      info["object"].(string),
```
Comment (vsukhin) [inline pkg/cloud/data/scraper/loader.go:55]: possible panic, if we change CheckStringKey method

File: pkg/cloud/data/scraper/loader.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		ExecutionID: info["executionId"].(string),
+		TestName:    info["testName"].(string),
+	}
+
+	if info["testSuiteName"] != nil {
+		if s, ok := info["testSuiteName"].(string); ok {
+			req.TestSuiteName = s
+		}
+	}
```
Comment (vsukhin) [inline pkg/cloud/data/scraper/loader.go:64]: guess, can be reused above method

File: pkg/executor/scraper/filesystem_extractor.go (hunk 1 above)
Comment (dejanzele) [reply pkg/executor/scraper/filesystem_extractor.go:38]: I wanted to build generic Extractor interface which should in theory be able to extract data from anywhere.
In our case, we only walk directories so we only have 1 implementation and we deal with files there.

File: pkg/executor/scraper/loader.go (hunk 2 above)
Comment (dejanzele) [reply pkg/executor/scraper/loader.go:8]: Or that, yeah, it extracts then uploads

File: pkg/executor/scraper/scraper.go (hunk 4 above)
Comment (dejanzele) [reply pkg/executor/scraper/scraper.go:13]: keep in mind I need to refactor every single executoro

**Pull request merged**